ofono/src/cell-info-control.c

```diff
@@ -0,0 +1,282 @@
+	g_free(self);
+}
+
+CellInfoControl *cell_info_control_get(const char* path)
+{
+	if (path) {
+		CellInfoControlObject *self = NULL;
+
+		if (cell_info_control_table) {
+			self = g_hash_table_lookup(cell_info_control_table,
+				path);
+		}
+		if (self) {
+			/* Already there */
+			return cell_info_control_ref(&self->pub);
+		} else {
+			/* Create a new one */
+			self = g_new0(CellInfoControlObject, 1);
+			self->pub.path = self->path = g_strdup(path);
+			self->refcount = 1;
+
+			/* Create the table if necessary */
+			if (!cell_info_control_table) {
+				cell_info_control_table =
+					g_hash_table_new(g_str_hash,
+						g_str_equal);
+			}
+			g_hash_table_insert(cell_info_control_table,
+				self->path, self);
+			DBG("%s created", path);
+			return &self->pub;
+		}
+	}
+	return NULL;
+}
+
+CellInfoControl *cell_info_control_ref(CellInfoControl *ctl)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self) {
+		self->refcount++;
+	}
+	return ctl;
+}
+
+void cell_info_control_unref(CellInfoControl *ctl)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self && !--self->refcount) {
+		cell_info_control_free(self);
+	}
+}
+
+void cell_info_control_set_cell_info(CellInfoControl *ctl,
+				struct ofono_cell_info *ci)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self && ctl->info != ci) {
+		ofono_cell_info_unref(ctl->info);
+		ctl->info = ofono_cell_info_ref(ci);
+		cell_info_control_update_all(self);
+	}
+}
+
+void cell_info_control_drop_all_requests(CellInfoControl *ctl)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self) {
+		cell_info_control_drop_all_requests_internal(self);
+		cell_info_control_update_all(self);
+	}
+}
+
+void cell_info_control_drop_requests(CellInfoControl *ctl, void *tag)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self && tag) {
+		if (self->enabled &&
+			g_hash_table_remove(self->enabled, tag) &&
+			!g_hash_table_size(self->enabled)) {
+			g_hash_table_unref(self->enabled);
+			self->enabled = NULL;
+			ofono_cell_info_set_enabled(ctl->info, FALSE);
+		}
+		if (self->set_interval &&
+			g_hash_table_remove(self->set_interval, tag)) {
+			int ms;
+
+			if (g_hash_table_size(self->set_interval)) {
+				ms = cell_info_control_get_interval(self);
+			} else {
+				g_hash_table_unref(self->set_interval);
+				self->set_interval = NULL;
+				ms = INT_MAX;
+			}
+			ofono_cell_info_set_update_interval(ctl->info, ms);
+		}
+	}
+}
+
+void cell_info_control_set_enabled(CellInfoControl *ctl,
+	void *tag, ofono_bool_t enabled)
+{
+	CellInfoControlObject *self = cell_info_control_object_cast(ctl);
+
+	if (self && tag) {
+		gboolean was_enabled = (self->enabled != NULL);
+		gboolean is_enabled;
+
+		if (enabled) {
+			if (!self->enabled) {
+				self->enabled = g_hash_table_new(g_direct_hash,
+					g_direct_equal);
+			}
+			g_hash_table_insert(self->enabled, tag, tag);
```

Comment:
Since we already require GLib 2.32, could use `g_hash_table_add()` to hide the implementation detail of key = value.

Reply:
Thank you, that `g_hash_table_add()` thing escaped my attention. Yeah, why not.